Cleave which does not ignore armor

AbitTheGray opened the pull request:
This is not ready for release!

Sven has 2 abilities with same values.
2nd (W) is his normal cleave
4th (D) is new version with cleave which does not ignore armor

This is just for comparison purpose.
More abilities (Kunkka and Magnus) will come ASAP. Then Battlefury and last talents (PA, Alchemist, Doom).

Next message here will be when all 3 abilities will be working (heroes will still have 2 versions of the ability).

For testing I recommend using `-createhero dragon_knight enemy` and `-levelbots 50`. Giving them full passive and talent for doubling it.
Also used `-spawncamps` multiple times.

AbitTheGray commented:
This version is still **__only for testing__**


Changes:
 - Cleave from target's position can be turned on by setting `unit.cleaveFromTarget = true` or `event.cleaveFromTarget = true` on attack (OnAttackLanded by `MODIFIER_EVENT_ON_ATTACK_LANDED`) or `modifier.cleaveFromTarget = true` on the modifier
 - Cleave no longer ignores armor
 - New abilities:
   - sven_great_cleave_oaa
   - kunkka_tidebringer_oaa
   - magnataur_empower_oaa
 - New abilities are added as 4th ability (may break x_marks_the_spot)


Usage:
```lua
cleaveInfo = {
    startRadius = 150,
    endRadius = 330,
    length = 625
}
damageMult = 1.0
damageType = DAMAGE_TYPE_PHYSICAL
--damageFlags = DOTA_DAMAGE_FLAG_IGNORES_PHYSICAL_ARMOR -- Original Cleave
damageFlags = DOTA_DAMAGE_FLAG_NONE -- OAA Cleave
--damageFlags = DOTA_DAMAGE_FLAG_REFLECTION -- If we do not want them to return damage from cleave

targetTeam = DOTA_UNIT_TARGET_TEAM_ENEMY
targetUnit = DOTA_UNIT_TARGET_BASIC | DOTA_UNIT_TARGET_HERO
targetFlags = DOTA_UNIT_TARGET_FLAG_NONE


-- Does not work for illusions (still play sound and show particle when provided)
-- Does not work on allies, towers, barracks, buildings and wards
-- Does not work with BREAK effect ( unit:PassivesDisabled() )
-- Does not work if event == nil or event.target == nil or event.target ~= self:GetParent()
false/{}/nil CDOTABaseAbility:PerformCleaveOnAttack(event, nil, nil, nil, nil, nil)
false/{}/nil CDOTABaseAbility:PerformCleaveOnAttack(event, damageMult, cleaveInfo, soundName, particleNameCleave, particleNameHit)

-- Shortcut to PerformCleave(...)
{}/nil CDOTA_BaseNPC:PerformCleave(nil, nil, damage, nil, nil, nil, nil, nil)
{}/nil CDOTA_BaseNPC:PerformCleave(originalTarget, cleaveInfo, damage, damageType, damageFlags, targetTeam, targetUnit, targetFlags)

-- Shortcut to PerformCleave(...)
{}/nil CDOTABaseAbility:PerformCleave(nil, nil, damage, nil, nil, nil, nil, nil)
{}/nil CDOTABaseAbility:PerformCleave(originalTarget, cleaveInfo, damage, damageType, damageFlags, targetTeam, targetUnit, targetFlags)

-- Does not check validity of target (not illusion, not passives disables...), just performs the cleave effect
{}/nil PerformCleave(attacker, nil, nil, nil, nil, nil, nil, damage, nil, nil, nil, nil, nil)
{}/nil PerformCleave(attacker, ability, originalTarget, teamNumber, position, direction, cleaveInfo, damage, damageType, damageFlags, targetTeam, targetUnit, targetFlags)
```


TODO:
 - [ ] Change Kunkka's Tidebringer
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Add particle to weapon when the ability is active
     - [ ] Support different weapons
   - [ ] Hide cleave buff
   - [ ] Only work when toggle is active
   - [ ] On first level-up of the skill, toggle it on
   - [ ] Replace old ability by new one
 - [ ] Change Sven's Great Cleave
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Hide cleave buff
   - [ ] Replace old ability by new one
 - [ ] Change Magnus's Empower
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Fix cleave damage percentage and bonus damage increase from talent
   - [ ] Replace old ability by new one
 - [ ] Change Battlefury
 - [ ] Change talents
   - [ ] Alchemist
   - [ ] PA
   - [ ] Doom
 - [ ] Clean-up and polish

AbitTheGray commented:
This version is still **__only for testing__**

Yesterday I was out of PC for about 10 hours and today I was taking it slow.
Here should be all

Changes:
 - Swapped 2nd and 3rd parameter of `CDOTABaseAbility:PerformCleaveOnAttack` to match other (custom) cleave functions (first `cleaveInfo` then `daamge`)
 - Completed implementation of Kunkka's Tidebringer
 - Added all tooltips (still require verification)

TODO:
 - [ ] Change Kunkka's Tidebringer
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Add support different weapon glow effects (based on the equiped item)
   - [ ] Find and use the right "Tidebringer Going off" sound
   - [ ] Replace old ability by new one
 - [ ] Change Sven's Great Cleave
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Replace old ability by new one
 - [ ] Change Magnus's Empower
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Fix cleave damage percentage and bonus damage increase from talent
     - [ ] Try to get the current one working
     - [ ] Reimplement the talent
   - [ ] Replace old ability by new one
 - [ ] Change Battlefury
 - [ ] Change talents
   - [ ] Alchemist
   - [ ] PA
   - [ ] Doom
 - [ ] Clean-up and polish
   - [ ] Round 2

DerpySoccerBall commented:
Great work abit, there is another ability you need to change which is tiny's tree grab. since this just got buffed to 100% cleave it would be nice to be fixed.

AbitTheGray commented:
@DerpySoccerBall Tiny's Tree Grab does not ignore armor (at least by Dota Wiki)
https://dota2.gamepedia.com/Cleave#Tree_Grab
"Unlike Cleave, Tree Grab's damage considers armor value as well, instead of only armor type. It also counts as spell damage, and thus fully works with spell lifesteal and spell damage amplification."

AbitTheGray commented:
This version is still **__only for testing__**

Changes:
 - Added new Battlefury as `item_bfury_oaa_1`,`item_bfury_oaa_2`,`item_bfury_oaa_3`,`item_bfury_oaa_4` and `item_bfury_oaa_5` (recipe: `item_recipe_bfury_oaa_X` `for X=1,6`)
   - Does not have `Quell`
   - Cannot destroy trees / wards / mines
   - Uses "new Cleave"
   - Copied values from old Battlefury

TODO:
 - [ ] Verify that cleave cannot hit wards/buildings and cannot cleave from them (using `PerformCleaveOnAttack(...)`)
 - [ ] Change Kunkka's Tidebringer
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Add support different weapon glow effects (based on the equiped item)
   - [ ] Find and use the right "Tidebringer Going off" sound
   - [ ] Replace old ability by new one
 - [ ] Change Sven's Great Cleave
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Replace old ability by new one
 - [ ] Change Magnus's Empower
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Fix cleave damage percentage and bonus damage increase from talent
     - [ ] Try to get the current one working
     - [ ] Reimplement the talent
   - [ ] Replace old ability by new one
 - [ ] Change Battlefury
   - [ ] Verify tooltips
   - [X] Perform cleave
     - [ ] Add cleave effect (end of `game/scripts/vscripts/items/bfury_oaa.lua`)
   - [ ] Add Quell (`quelling_bonus`, `quelling_bonus_ranged`)
   - [ ] Add `OnSpellStart()` to destroy tree or other (wards, mines)
     - [ ] Fix different range for trees and wards/mines
     - [ ] Destroy trees or wards/mines
     - [ ] Figure out how to detect trees
   - [ ] Change recipes to primary show previous version of Battlefury instead of Dezzolator
   - [ ] Replace old BFs by new one (reference new `item_bfury_1`)
 - [ ] Change talents
   - [ ] Alchemist
   - [ ] PA
   - [ ] Doom
 - [ ] Clean-up and polish
   - [ ] Round 2

AbitTheGray commented:
This version is still **__only for testing__**

Changes:
 - Added new Generic Cleave talents as `special_bonus_cleave_oaa_5`, `special_bonus_cleave_oaa_15`, `special_bonus_cleave_oaa_20`, `special_bonus_cleave_oaa_25`, `special_bonus_cleave_oaa_30`, `special_bonus_cleave_oaa_60`, `special_bonus_cleave_oaa_100` and `special_bonus_cleave_oaa_150`
   - Implemented by `modifier_cleave_talents_oaa` positioned at {0,0,0} with no entity (triggering at `MODIFIER_EVENT_ON_ATTACK_LANDED` and then checking for all OAA cleave talents mentioned above)

TODO:
 - [ ] Verify that cleave cannot hit wards/buildings and cannot cleave from them (using `PerformCleaveOnAttack(...)`)
 - [ ] Change Kunkka's Tidebringer
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Add support different weapon glow effects (based on the equiped item)
   - [ ] Find and use the right "Tidebringer Going off" sound
   - [ ] Replace old ability by new one
 - [ ] Change Sven's Great Cleave
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Replace old ability by new one
 - [ ] Change Magnus's Empower
   - [ ] Verify cleave + hit particles
   - [ ] Verify tooltips
   - [ ] Fix cleave damage percentage and bonus damage increase from talent
     - [ ] Try to get the current one working
     - [ ] Reimplement the talent
   - [ ] Replace old ability by new one
 - [ ] Change Battlefury
   - [ ] Verify tooltips
   - [X] Perform cleave
     - [ ] Add cleave effect (end of `game/scripts/vscripts/items/bfury_oaa.lua`)
   - [ ] Add Quell (`quelling_bonus`, `quelling_bonus_ranged`)
   - [ ] Add `OnSpellStart()` to destroy tree or other (wards, mines)
     - [ ] Fix different range for trees and wards/mines
     - [ ] Destroy trees or wards/mines
     - [ ] Figure out how to detect trees
   - [ ] Verify the item is not giving stats to non-owner
   - [ ] Change shop to primary show previous version of Battlefury instead of Dezzolator
   - [ ] Disable old Battlefury
 - [X] Add generic cleave talents (5, 15, 20, 25, 30, 60, 100, 150 percent cleave)
   - [X] Add global modifier `modifier_cleave_talents_oaa` to position {0, 0, 0} which triggers at all attacks and checks for cleave talents (uses highest cleave talent)
     - [ ] Modify it to perform cleave for each talent
   - [ ] React on level-up and level-down of the talent to add/remove modifier which grants cleave
 - [ ] Change ability of Cleaver boss
 - [ ] Clean-up and polish

AbitTheGray commented:
This version is still **__only for testing__**

Changes:
 - Fixed and completed all 3 abilities
 - Added support for cleave to emmit sound on every hit target
 - Fixed Kunkka's particle on weapon
 - Replaced all old abilities/items with new ones
 - Reworked how cleave talents are used

Notes:
 - Kunkka's blade's effect does not react on refresh of the ability. Ability itself it not affected by this "bug" (but going-off-cooldown sound is affected)
 - Big thanks to Darkonius for pointing me in right direction and Warpdragon for... being there

TODO:
 - [ ] Verify that cleave cannot hit wards/buildings and cannot cleave from them
 - [ ] Change Kunkka's Tidebringer
   - [ ] Verify tooltips
   - [ ] Check whenever Rubick cannot steal the ability but Morphling can morph it
 - [ ] Change Sven's Great Cleave
   - [ ] Verify tooltips
 - [ ] Change Magnus's Empower
   - [ ] Verify tooltips
   - [ ] Fix cleave damage percentage and bonus damage increase from talent
 - [ ] Change Battlefury
   - [ ] Verify tooltips
   - [ ] Add Quell (`quelling_bonus`, `quelling_bonus_ranged`)
   - [ ] Add `OnSpellStart()` to destroy tree or other (wards, mines)
 - [ ] Change ability of Cleaver boss
 - [ ] Clean-up and polish

Trildar commented on game/scripts/vscripts/abilities/oaa_kunkka_tidebringer.lua:
```diff
@@ -0,0 +1,236 @@
+end
+
+function modifier_kunkka_tidebringer_oaa_passive:GetAttributes()
+  return bit.bor(MODIFIER_ATTRIBUTE_IGNORE_INVULNERABLE, MODIFIER_ATTRIBUTE_PERMANENT)
```
In my experience, both of these are usually unnecessary. What you should probably have is the `RemoveOnDeath` method returning false.

AbitTheGray replied:
Better safe then sorry.

I added the `RemoveOnDeath` but same argument can be used for `IsPermanent`.

Trildar commented on game/scripts/vscripts/abilities/oaa_kunkka_tidebringer.lua:
```diff
@@ -0,0 +1,236 @@
+end
+
+function modifier_kunkka_tidebringer_oaa_passive:GetModifierPreAttack_BonusDamage( event )
+  if self:GetParent() ~= event.attacker then
```
This is not how modifier properties work. You just need to return a value, and it applies to the unit that the modifier is on. Only events, typically anything with `On` in the method name, need this sort of check.

Trildar commented on game/scripts/vscripts/abilities/oaa_kunkka_tidebringer.lua:
```diff
@@ -0,0 +1,236 @@
+  local ability = self:GetAbility()
+
+  -- Toggled off
+  if ( not ability:GetAutoCastState() ) or ( not ability:IsCooldownReady() ) then
```
You're missing handling for manual casts of the ability. Refer to Glaives of Wisdom or Arcane Orb for that.

AbitTheGray replied:
I am aware of that.
I still need to verify Magnus's Empower talent which did not work before or Bfury's active.
All 3 of those had lowest priority for me because it works without it.

Trildar commented on game/scripts/vscripts/abilities/oaa_kunkka_tidebringer.lua:
```diff
@@ -0,0 +1,236 @@
+end
+
+function modifier_kunkka_tidebringer_oaa_cooldown:GetAttributes()
+  return bit.bor(MODIFIER_ATTRIBUTE_IGNORE_INVULNERABLE, MODIFIER_ATTRIBUTE_PERMANENT)
```
Seems weird to me to give a modifier the permanent attribute when it's not actually permanent. And again, you probably want `RemoveOnDeath` false.

AbitTheGray replied:
I gave it permanent attribute to not be removed by some weird Dota internal mechanic.

AbitTheGray commented:
This version is still **__only for testing__**.

Changes:
 - Another change to Cleave talents
 - A lot of fixes

TODO:
 - [ ] Verify that cleave cannot hit wards/buildings and cannot cleave from them
 - [ ] Change Kunkka's Tidebringer
   - [ ] Verify tooltips
   - [X] Add possibility to use it as active
   - [ ] Check whenever Rubick cannot steal the ability but Morphling can morph it
   - [ ] Cannot target allied units, but applies the bonus damage when attacking while on autocast and off cooldown, without cleaving or going on cooldown.
   - [ ] Verify that illusions does not benefit from damage (they play sound but not cleave already).
 - [ ] Change Sven's Great Cleave
   - [ ] Verify tooltips
 - [ ] Change Magnus's Empower
   - [ ] Verify tooltips
   - [ ] Fix cleave damage percentage and bonus damage increase from talent
 - [ ] Change Battlefury
   - [ ] Verify tooltips
   - [ ] Add Quell (`quelling_bonus`, `quelling_bonus_ranged`)
   - [ ] Add `OnSpellStart()` to destroy tree or other (wards, mines)
 - [ ] Change ability of Cleaver boss
 - [ ] Clean-up and polish

AbitTheGray commented:
Changes:
 - Fixed and verified abilities
 - Fixed tooltips
 - Replaced `\\n\\n` by `<br/><br/>` which works the same but does not leave `\` before the new line (when space char is not in front of it)

Note:
 - Phantom Assassin will always cleave from targets location (from both her talent and Battlefury) because of her Q not teleporting her in front of the target (and I could not find a way to detect it)

Not done, do?
 - Change ability of Cleaver boss (`boss_great_cleave`)
 - Add `OnSpellStart()` of Battlefury to destroy tree or other (wards, mines)
 - Add Quell (`quelling_bonus`, `quelling_bonus_ranged`) to Battlefury
 - Fix Empower's Modifier Client-side Tooltip = add talents to `npc_abilities_custom.txt` or `npc_abilities_override.txt` for `C_DOTA_BaseNPC:FindTalentValue`

AbitTheGray commented:
Just did testing and found no problem with it.
Thanks ti Darkonius for mentioning Spell Immunity on which I forgot.

DarkoniusXNG commented:
You are welcome Abit. Just one more thing: Are particles for Sven Great Cleave displaying properly?
